File: project/drivers/driver/r/tests/testthat/test_sql.R
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

test_that("normalize positional params", {
  sql <- "SELECT * FROM t WHERE id = ? AND name = ?"
  out <- sb_normalize(sql, list(42, "Ada"))
  expect_equal(out$sql, "SELECT * FROM t WHERE id = $1 AND name = $2")
  expect_equal(out$params, list(42, "Ada"))
})

test_that("normalize named params", {
  sql <- "SELECT * FROM users WHERE name = @name AND active = :active"
  out <- sb_normalize(sql, list(name = "Ada", active = TRUE))
  expect_equal(out$sql, "SELECT * FROM users WHERE name = $1 AND active = $2")
  expect_equal(out$params, list("Ada", TRUE))
})

test_that("normalize binary params", {
  sql <- "SELECT ?"
  out <- sb_normalize(sql, list(as.raw(c(0x01, 0x02))))
  expect_equal(out$sql, "SELECT $1")
  expect_equal(out$params, list(as.raw(c(0x01, 0x02))))
})

test_that("statement chunker matches cross-driver conformance fixture", {
  skip_if_not_installed("jsonlite")
  # Locate the shared oracle fixture relative to the R driver root. The test
  # may run from tests/testthat (testthat) or the package root (R CMD check).
  candidates <- c(
    "../../../../../tests/conformance/drivers/chunker_conformance/cases.json",
    "../../../../tests/conformance/drivers/chunker_conformance/cases.json",
    file.path(
      Sys.getenv("SCRATCHBIRD_REPO_ROOT", unset = "."),
      "project/tests/conformance/drivers/chunker_conformance/cases.json"
    )
  )
  cases_path <- Filter(file.exists, candidates)
  skip_if(length(cases_path) == 0, "chunker conformance cases.json not found")
  fixture <- jsonlite::fromJSON(cases_path[[1]], simplifyDataFrame = FALSE)
  for (case in fixture$cases) {
    expected <- as.character(unlist(case$expected))
    actual <- as.character(split_top_level_statements(case$input))
    expect_identical(actual, expected, info = case$name)
  }
})

File: packaging/2026.07.03/drivers/ruby/examples/test/test_sql.rb
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

require "test_helper"

class TestSql < Minitest::Test
  def test_normalize_positional
    sql = "SELECT * FROM t WHERE id = ? AND name = ?"
    normalized = Scratchbird::Sql.normalize(sql, [42, "Ada"])
    assert_equal "SELECT * FROM t WHERE id = $1 AND name = $2", normalized.sql
    assert_equal [42, "Ada"], normalized.params
  end

  def test_normalize_named
    sql = "SELECT * FROM users WHERE name = @name AND active = :active"
    normalized = Scratchbird::Sql.normalize(sql, { name: "Ada", active: true })
    assert_equal "SELECT * FROM users WHERE name = $1 AND active = $2", normalized.sql
    assert_equal ["Ada", true], normalized.params
  end

  def test_normalize_binary
    sql = "SELECT ?"
    normalized = Scratchbird::Sql.normalize(sql, ["\x01\x02".b])
    assert_equal "SELECT $1", normalized.sql
    assert_equal ["\x01\x02".b], normalized.params
  end

  def test_normalize_callable_escape_syntax
    normalized = Scratchbird::Sql.normalize_callable("{ ? = call abs(?) }", [-3])
    assert_equal "select abs($1) as return_value", normalized.sql
    assert_equal [-3], normalized.params
  end

  def test_normalize_callable_sql_passthrough
    sql = Scratchbird::Sql.normalize_callable_sql("SELECT 1")
    assert_equal "SELECT 1", sql
  end
end
